src/protocols/secure_channel/CASEServer.cpp

```diff
@@ -46,76 +46,51 @@ CHIP_ERROR CASEServer::ListenForSessionEstablishment(Messaging::ExchangeManager
     mExchangeManager          = exchangeManager;
     mGroupDataProvider        = responderGroupDataProvider;
 
-    Cleanup();
+    ChipLogProgress(Inet, "CASE Server enabling CASE session setups");
+    mExchangeManager->RegisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::CASE_Sigma1, this);
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR CASEServer::InitCASEHandshake(Messaging::ExchangeContext * ec)
+CHIP_ERROR CASEServer::OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, Messaging::ExchangeDelegate *& newDelegate)
 {
-    ReturnErrorCodeIf(ec == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    if (mPairingSession.HasValue())
+        return CHIP_ERROR_NO_MEMORY;
+
+    CASESession * session = &mPairingSession.Emplace();
 
+    ChipLogProgress(Inet, "CASE Server allocated pairing: %p", session);
     // Setup CASE state machine using the credentials for the current fabric.
-    GetSession().SetGroupDataProvider(mGroupDataProvider);
+    session->SetGroupDataProvider(mGroupDataProvider);
     ReturnErrorOnFailure(
-        GetSession().ListenForSessionEstablishment(*mSessionManager, mFabrics, mSessionResumptionStorage, this,
-                                                   Optional<ReliableMessageProtocolConfig>::Value(GetLocalMRPConfig())));
-
-    // Hand over the exchange context to the CASE session.
-    ec->SetDelegate(&GetSession());
+        session->ListenForSessionEstablishment(*mSessionManager, mFabrics, mSessionResumptionStorage, this,
+                                               Optional<ReliableMessageProtocolConfig>::Value(GetLocalMRPConfig())));
+    newDelegate = session;
```

Comment:
This problem already exists before this PR, thanks for pointing it out.

Reply:
Before this PR we didn't _need_ to handle OnExchangeClosing.  We didn't change any of our state until `CASEServer::OnMessageReceived`, so if the exchange got closed before that point it wasn't a problem: the exchange would get closed, we would still be listening for incoming Sigma1, and everything would be fine.

But this PR changes the logic so that we now change our state much earlier, and then don't undo the state change on failure, ending up without the ability to accept incoming CASE connections.

Reply:
So do you take consideration that the exchange may closed in middle of any state, then the `CASESession` object will be leaked.

Reply:
What would close the exchange "in the middle of any state", exactly, for the specific case of this exchange?